src/Dialogs/Weather/SkysightDialog.cpp

```diff
@@ -0,0 +1,284 @@
+#include "Interface.hpp"
+
+#include "Language/Language.hpp"
+#include "LocalPath.hpp"
+#include "Net/HTTP/Session.hpp"
+
+#include "MapWindow/OverlayBitmap.hpp"
+#include "MapWindow/GlueMapWindow.hpp"
+#include "Dialogs/Error.hpp"
+
+#include "Util/StaticString.hxx"
+#include "Language/Language.hpp"
+#include "Time/BrokenDateTime.hpp"
+#include "Formatter/TimeFormatter.hpp"
+#include "Formatter/LocalTimeFormatter.hpp"
+
+#include "Weather/Skysight/Skysight.hpp"
+
+class SkysightListItemRenderer
+{
+  TwoTextRowsRenderer row_renderer; 
+
+public:
+  SkysightListItemRenderer(): skysight(DataGlobals::GetSkysight()) {}
+
+  void Draw(Canvas &canvas, const PixelRect rc, unsigned i); 
+
+  unsigned CalculateLayout(const DialogLook &look) {
+    return row_renderer.CalculateLayout(*look.list.font_bold, look.small_font);
+  }
+
+private:
+  std::shared_ptr<Skysight> skysight;
+};
+
+/**
+ * RENDERER FOR METRICS POPUP LIST
+ */
+class SkysightMetricsListItemRenderer: public ListItemRenderer
+{
+  TextRowRenderer row_renderer;
+  std::shared_ptr<Skysight> skysight;
+
+public:
+  SkysightMetricsListItemRenderer(): skysight(DataGlobals::GetSkysight()) {}
+  
+  unsigned CalculateLayout(const DialogLook &look) {
+    return row_renderer.CalculateLayout(*look.list.font);
+  }
+
+  virtual void OnPaintItem(Canvas &canvas, const PixelRect rc, unsigned i) override {
+    row_renderer.DrawTextRow(canvas, rc, skysight->GetMetric(i).name.c_str());
+  }
+  
+  static const TCHAR* HelpCallback(unsigned i)
+  {
+    std::shared_ptr<Skysight> skysight = DataGlobals::GetSkysight();
+
+    if (!skysight)
+      return _("No description available.");
+
+    tstring helptext = skysight->GetMetric(i).desc;
+    TCHAR *help = new TCHAR[helptext.length() + 1];
+    std::strcpy(help, helptext.c_str());
+    return help;
+  }  
+};
+
+class SkysightWidget final
+  : public ListWidget, private ActionListener, private Timer
+{
+  enum Buttons {
+    UPDATE,
+    UPDATE_ALL
+  };
+
+  ButtonPanelWidget *buttons_widget;
+
+  Button *update_button, *updateall_button;
+
+  struct ListItem {
+    StaticString<255> name;
+
+    gcc_pure bool operator<(const ListItem &i2) const
+    {
+      return StringCollate(name, i2.name) < 0;
+    }
+  };
+
+  SkysightListItemRenderer row_renderer;
+  std::shared_ptr<Skysight> skysight;
+
+public:
+  explicit SkysightWidget(std::shared_ptr<Skysight> &&_skysight)
+    : skysight(std::move(_skysight)) {}
+
+  void SetButtonPanel(ButtonPanelWidget &_buttons) {
+    buttons_widget = &_buttons;
+  }
+
+  void CreateButtons(ButtonPanel &buttons);
+
+private:
+  void UpdateList();
+  void UpdateClicked();
+  void UpdateAllClicked();
+
+public:
+  /* virtual methods from class Widget */
+  virtual void Prepare(ContainerWindow &parent,
+                       const PixelRect &rc) override;
+  virtual void Unprepare() override;
+
+protected:
+  /* virtual methods from ListItemRenderer */
+  virtual void OnPaintItem(Canvas &canvas, const PixelRect rc,
```

Comment:
Redundant `virtual`

Reply:
remove virtual, i.e. virtual void  -> void.

Are you able to explain why it is redundant? I am not an expert of c++. I can remove virtual from all function in the class.